src/tactic/polyrith.lean

```diff
@@ -275,34 +276,44 @@ ch '('
     <|> sub_parser poly_parser <|> mul_parser poly_parser <|> pow_parser poly_parser)
   <* ch ')'
 
+meta instance : non_null_json_serializable poly :=
+{ to_json := λ p, json.null,  -- we don't actually need this, but the typeclass asks for it
+  of_json := λ j, do
+    s ← of_json string j,
+    match poly_parser.run_string s with
+    | sum.inl s := exceptional.fail format!"unable to parse polynomial from.\n\n{s}"
+    | sum.inr p := pure p
+    end}
+
+/-- A schema for success messages from the python script -/
+@[derive [non_null_json_serializable, inhabited]]
+structure sage_json_success :=
+(success : {b : bool // b = tt})
+(trace : option string := none)
+(data : option (list poly) := none)
+
+/-- A schema for failure messages from the python script -/
+@[derive [non_null_json_serializable, inhabited]]
+structure sage_json_failure :=
+(success : {b : bool // b = ff})
+(error_name : string)
+(error_value : string)
+
 /-- Parse the json output from `scripts/polyrith.py` into either an error message, a list of `poly`
 objects, or `none` if only trace output was requested. -/
 meta def convert_sage_output (j : json) : tactic (option (list poly)) :=
 do
-  json.object obj ← pure j | fail!"Must be an object",
-  let obj := rbmap.from_list obj,
-  json.of_bool success ← obj.find "success" | fail!"internal error: missing success field",
-  if success then do
-    do
-    { some t ← pure (obj.find "trace") | skip,
-      json.of_string t ← pure t | fail!"internal error: trace must be a string",
-      tactic.trace t },
-    do
-    { some d ← pure (obj.find "data") | pure none,
-      json.array l ← some d | fail!"internal error: data field must be a string",
-      l ← l.mmap $ λ x, do
-      { json.of_string poly_s ← pure x | fail!"internal error: entries must be strings",
-        pure poly_s },
-      l ← l.mmap $ λ x, match poly_parser.run_string x with
-        | sum.inl s := fail!"internal error: unable to parse polynomial from.\n\n{s}"
-        | sum.inr p := pure p
-        end,
-      pure l }
-  else do
-    json.of_string kind ← obj.find "error_name",
-    json.of_string message ← obj.find "error_value",
-    fail!"polyrith failed to retrieve a solution from Sage! {kind}: {message}"
-
+  let e : exceptional (sage_json_success ⊕ sage_json_failure) :=
+    -- try the error format first, so that if both fail we get the message from the success parser
+    (sum.inr <$> of_json sage_json_failure j) <|> (sum.inl <$> of_json sage_json_success j),
+  match e with
+  | exceptional.exception f := exceptional.exception (λ s, format!"internal json error: " ++ f s)
+  | exceptional.success (sum.inr f) :=
+      fail!"polyrith failed to retrieve a solution from Sage! {f.error_name}: {f.error_value}"
+  | exceptional.success (sum.inl s) := do
+      do { some t ← pure s.trace | skip, tactic.trace t},
+      pure s.data
+  end
```

Comment:
```suggestion
(do f ← of_json sage_json_failure j,
    fail!"polyrith failed to retrieve a solution from Sage! {f.error_name}: {f.error_value}") <|>
(do s ← decorate_ex "internal json error: " $ of_json sage_json_success j,
    s.trace.mmap trace,
    pure s.data)
```
I think this is a slicker version of the same thing!

Reply:
That doesn't work correctly, because if an error message is recieved the failure is treated as a parsing failure, and then it tries to reparse as a success message. Still a helpful comment though, I didn't know about `decorate_ex` and didn't think of `option.mmap`!